packages/mcdoc/src/type/index.ts

```diff
@@ -158,6 +159,12 @@ export interface ConcreteType extends McdocBaseType {
 	typeArgs: McdocType[]
 }
 
+export interface MappedType extends McdocBaseType {
+	kind: 'mapped'
+	child: McdocType
+	mapping: { [path: string]: SimplifiedMcdocType }
+}
```

Comment:
I am trying this right now, but it ends up infecting everything with `MappedType`. Because `checkShallow` uses the nested types of lists and structs to set to set as `childDefinitions`.

Reply:
Though I agree that importing `SimplifiedMcdocType` is not very nice. Could this be cleaned up by moving the simplified fields into `mcdoc/src/type/simplified.ts`?

Reply:
I kind of disagree with that. Simplify only makes sense with a runtime value, so it - and by extension also simplified types - belong in runtime imo.

I also don't really like the idea of crating a mapped type here, as a mapped type can never actually appear in any type tree, it's just an artifact / implementation detail of simplify.

There has to be a cleaner way to do this...


Would it for example be possible to return the `mapping` from simplify instead, and then also from checkShallow as part of the `ValidDefintionResult`?

The checker tree could then store this info and pass it back into checkShallow when simplifying the children.

Reply:
> Would it for example be possible to return the mapping from simplify instead, and then also from checkShallow as part of the ValidDefintionResult?
> The checker tree could then store this info and pass it back into checkShallow when checking the children.

I had considered this, but the issue is that other places (for example the completer) need to be able to work with these simplified types (which could contain references to template params, but without the original type args information).

Reply:
> Simplify [`SimplifiedMcdocType`] only makes sense with a runtime value

This might be true semantically in the runtime checker, but looking purely at the types, they're just a subset of `McdocType`, a more restricted kind of type.

Reply:
I have moved the simplified types to their own file, I hope this removes a bit of the curseness.

Reply:
> > Simplify [`SimplifiedMcdocType`] only makes sense with a runtime value
> 
> This might be true semantically in the runtime checker, but looking purely at the types, they're just a subset of `McdocType`, a more restricted kind of type.

Eh, I would think it makes the most sense to have simplify and simplified types in the same file. Looking at simplify makes it clear.

Reply:
> > Would it for example be possible to return the mapping from simplify instead, and then also from checkShallow as part of the ValidDefintionResult?
> > The checker tree could then store this info and pass it back into checkShallow when checking the children.
> 
> I had considered this, but the issue is that other places (for example the completer) need to be able to work with these simplified types (which could contain references to template params, but without the original type args information).

How about just allowing to put mapping info into any simplified type (or even just storing them as attribute)?

Additional properties don't change assignability (you might just lose some info when an McdocType is expected and not a simplified one directly, but if it's an attribute technically everyone could read it), but I don't immediately see why that would be an issue.

Type mapping info really only should be relevant when a simplified type is expected, right?

You would still need to explicitly pass the type mappings from the parent definition when resolving the child, but I don't think that's a big deal.

Reply:
> Essentially now every consumer of McdocType needs to handle mapped type, even though it cannot appear in the tree normally.

The runtime checker will attach type info to nodes as `SimplifiedMcdocType`, so almost no mcdoc consumer will use `McdocType` directly.

> Having simplified types in a place where you can't create them doesn't make sense to me, and having a mapped type in the main tree which is implementation detail from simplify also doesn't seem very nice.

The normal `McdocType`s are also in a place where you can't create them directly. They're only created by the mcdoc binder. I see `MappedType` as an internal type, just because it doesn't naturally it's still a valid mcdoc type.

> How about just allowing to put mapping info into any simplified type (or even just storing them as attribute)?

I don't like this solution, as it's very easy to forget about them. With attributes if you ignore them you get semi good results. If you ignore the type mappings, the references won't be found and will result in an empty union.

Reply:
> The normal `McdocType`s are also in a place where you can't create them directly. They're only created by the mcdoc binder. I see `MappedType` as an internal type, just because it doesn't occur naturally in mdoc source files, it's still a valid mcdoc type.

Binder and types are on the same level in my view. Runtime is kind of its own thing.



> > How about just allowing to put mapping info into any simplified type (or even just storing them as attribute)?
> 
> I don't like this solution, as it's very easy to forget about them. With attributes if you ignore them you get semi good results. If you ignore the type mappings, the references won't be found and will result in an empty union.

This is why imo simplify and simplified types should be in the root of runtime, and the only function resolving those references should be the simplify function. I don't see a point in implementing this logic multiple times, and have multiple functions deal with mappings and references. Imo they should be handled solely by simplify. I would prefer this solution _a lot_.

Maybe we should get a third opinion. @SPGoding ?

Reply:
I think I found a good compromise, check it out @NeunEinser